lib/mix/lib/mix/dep.ex

```diff
@@ -160,26 +204,16 @@ defmodule Mix.Dep do
     Enum.filter(all_deps, &(&1.app in apps))
   end
 
-  defp get_deps_with_children(all_deps, apps) do
-    deps = get_children(all_deps, apps)
-    apps = deps |> Enum.map(& &1.app) |> Enum.uniq()
-    get_deps(all_deps, apps)
-  end
-
-  defp get_children(_all_deps, []), do: []
-
-  defp get_children(all_deps, apps) do
-    # Current deps
-    deps = get_deps(all_deps, apps)
+  defp get_children(_all_deps, _seen, []), do: []
 
-    # Children apps
-    apps =
-      for %{deps: children} <- deps,
+  defp get_children(all_deps, seen, apps) do
+    children_apps =
+      for %{deps: children} <- get_deps(all_deps, apps),
           %{app: app} <- children,
+          app not in seen,
           do: app
 
-    # Current deps + children deps
-    deps ++ get_children(all_deps, apps)
+    apps ++ get_children(all_deps, children_apps ++ seen, children_apps)
```

Comment:
Seeing the `children_apps ++ seen` and then the `seen` check above, I wonder if this shouldn't be a set instead.